caas/kubernetes/provider/application/constraints.go

```diff
@@ -266,35 +281,131 @@ func processPodAffinity(pod *core.PodSpec, affinityLabels map[string]string) err
 	return nil
 }
 
-func configureConstraint(pod *core.PodSpec, resourceName core.ResourceName, value string) (err error) {
+func configureCharmConstraint(pod *core.PodSpec, resourceName core.ResourceName, requestValue, limitValue string) (err error) {
 	if len(pod.Containers) == 0 {
 		return nil
 	}
-	pod.Containers[0].Resources.Requests, err = MergeConstraint(resourceName, value, pod.Containers[0].Resources.Requests)
+	requestResourceQty, err := resource.ParseQuantity(requestValue)
+	if err != nil {
+		return errors.Annotatef(err, "invalid constraint value %q for %q", requestValue, resourceName)
+	}
+	limitResourceQty, err := resource.ParseQuantity(limitValue)
 	if err != nil {
-		return errors.Annotatef(err, "merging request constraint %s=%s", resourceName, value)
+		return errors.Annotatef(err, "invalid constraint value %q for %q", requestValue, resourceName)
+	}
+	for i, container := range pod.Containers {
+		if container.Name != constants.ApplicationCharmContainer {
+			continue
+		}
+
+		if pod.Containers[i].Resources.Requests, err = MergeConstraint(resourceName, requestResourceQty, pod.Containers[i].Resources.Requests); err != nil {
+			return errors.Annotatef(err, "merging request constraint %s=%s", resourceName, requestValue)
+		}
+		// If limits on charm containers are already set by user, we don't override them.
+		if pod.Containers[i].Resources.Limits != nil {
+			break
+		}
+		if pod.Containers[i].Resources.Limits, err = MergeConstraint(resourceName, limitResourceQty, pod.Containers[i].Resources.Limits); err != nil {
+			return errors.Annotatef(err, "merging limit constraint %s=%s", resourceName, limitValue)
+		}
+		break
 	}
-	for i := range pod.Containers {
-		pod.Containers[i].Resources.Limits, err = MergeConstraint(resourceName, value, pod.Containers[i].Resources.Limits)
-		if err != nil {
+	return nil
+}
+
+func DivideAndSpread(total uint64, parts int) []uint64 {
+	if parts <= 0 {
+		return nil
+	}
+
+	result := make([]uint64, parts)
+	quotient := total / uint64(parts)
+	remainder := total % uint64(parts)
+
+	for i := 0; i < parts; i++ {
+		if uint64(i) < remainder {
+			result[i] = quotient + 1
+		} else {
+			result[i] = quotient
+		}
+	}
+	return result
+}
+
+// configureWorkloadConstraint configures the workload container constraints and the limits for the charm container.
```

Comment:
Given the expanded behaviour, can we add extra detail to the comment

Reply:
sure added an example :)